Review Resolution Convention Spec (#5)
iotmodels/specs: resolution-convention.md

# Resolution Convention Spec

A model is composed by a set of interfaces identified by a unique id in the form of a DTMI.

The model repo stores one interface version per file following the structure defined below.

## Folder and File structure

A DTMI can be translated to a relative path with the following rules:

- Folder name is based on the DTMI
  - Convert to lower case invariant
  - Create a folder structure with a subfolder for each DTMI segment (separated by `:`)
- File name is the last DTMI segment plus the version number with the `.json` extension

```text
dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
```

Folders or files are associated to `CODEOWNERS` (Spec TBD) to ensure that any future updates are reviewed by the user who initially created the folder.

> >Note: Once a DTMI is registered with a specific casing all other case variants will be stored under the same folder owned by the same user.

Reviewer: This means that dtmi:com:example:..., dtmi:com:EXAMPLE:..., and dtmi:com:eXaMpLe:... must all have the same owner. Is that okay?

Author: yes

Reviewer: Seems reasonable. I consider this comment resolved.

### Casing

> Based on the file/folder rules, only one DTMI can be stored with specific casing.

Reviewer: Is the intention that someone with the DTMI (say from a P&P device) can use the DTMI to find the folder/file that contains the definition (without using an index file)? If so, then I think we might have problems because dtmi:com:example:Thermostat;1 and dtmi:com:example:thermostat;1 will map to the same file, which is not correct.

If there is a requirement that the DTMI must first be looked up in an index and only if there is a match then mapped to folder, then this might be okay. It would mean that dtmi:com:example:Thermostat;1 and dtmi:com:example:thermostat;1 cannot both be stored in the model repo even though both are valid and different DTMIs and model definitions. It would mean that the model repo cannot store all possible models, but only a subset of all possible models.

Author: we've removed the index

The model can keep casing in their IDs, but the mapping to the url will always be case insensitive. Case diffs can be set in different versions, all owned by the same user.

>dtmi:com:example:Thermostat;1 and dtmi:com:example:thermostat;1 will map to the same file, which is not correct

The repo will guarantee there will be only one of these two. The upcoming "repo sdk" will make sure that the retrieved ID is what was requested, with casing.

Reviewer: With this approach, customers will not always get what they expect.

P&P device A connects to IoT Hub and specifies dtmi:com:example:Thermostat;1. A solution maps that to dtmi/com/example/thermostat-1.json. Everything looks good.

P&P device B connects to IoT Hub and specifies dtmi:com:example:thermostat;1 (a different DTMI which is a different interface definition). A solution maps that to dtmi/com/example/thermostat-1.json. The solution gets the wrong interface definition for P&P device B.

Or are you saying that the solution will fail to resolve device B's DTMI because the case does not match the DTMI casing in the file?

Reviewer 2: This is not strictly a limitation of the repo, it is possible to use casing in both git/github and in URLs to allow differentiation between Thermostat and thermostat and ThermOsTat. I think the question is whether that provides a sensible user experience. To me it seems confusing. I could create a c# package/namespace called microSoft.aZure.digitalTwins.pArser and it would work, but it is confusing and we shouldn't allow people to do it.

Reviewer 3: > Or are you saying that the solution will fail to resolve device B's DTMI because the case does not match the DTMI casing in the file?

the resolver will find the url, download the json and see the @id is different, then it will fail resolution

Reviewer: Got it, so there is an additional check required: that the DTMI in the downloaded model exactly matches the DTMI from the P&P device.

I agree that unexpected casing is confusing, but sometimes developers use casing with purpose. In C#, casing is used as part of naming conventions. My biggest concern was that, unknowingly, someone would get the wrong model definition rather than getting none or an error. It sounds like the documentation and guidance will be to perform a DTMI comparison after download the file, so if a DTMI comes through with differing case, an error will be raised so the problem can addressed by the developer.

Reviewer: As long as we're very clear that the DTMI of the downloaded model must be compared using DTMI comparison rules against the source DTMI (and encode this in any libraries/tools we provide) then I think this is reasonable.

I consider this comment resolved.

> Based on the file/folder rules, only one DTMI can be stored with specific casing.

Reviewer 4: I think the wording `with specific casing` should be changed.
```suggestion
Based on the file/folder rules, only one DTMI can be stored regardless of id casing.
```

:heavy_check_mark: Valid

```text
dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
dtmi:com:example:thermostat;2 -> dtmi/com/example/thermostat-2.json
```

:x: Invalid

```text
dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
dtmi:com:example:thermostat;1 -> dtmi/com/example/thermostat-1.json
```

### Relative Paths

Each IoT Model Repo will be accessible at a URL, where all paths are relative to the location of base repo URL.

For a model repo at `https://iotmodels.github.io/registry` the referenced interfaces will be available from:

```rest
https://iotmodels.github.io/registry/dtmi/com/example/thermostat-1.json
```

The same folder structure can be used locally with local file paths

```rest
C:\localModelRepo\dtmi\com\example\thermostat-1.json
```

## Unique `@id` elements

The model repo must guarantee that all `@id` included in the interfaces are unique accross the repo to avoid collisions when loading different models.

> - `@id` elements from interfaces are guaranteed to be unique by the `model-index.json` file.

Reviewer: When DTDL, in a future version, supports reusable definitions other than interfaces, how will the model repo be updated to handle that?

Author: I dont see any problem by enabling more DTDL elements, like schemas with complex objects, to be published under the same convention. As long the parser uses the same DTMI resolver pattern sending those IDs in the callback.

Reviewer: Yeah, I expect that if/when we introduce reusable elements, it'll only be supported in DTDL vX (not DTDL v2), so we wouldn't expect to go back to DTDL v2 models and enable reuse for elements that have already been uploaded.

I don't see a problem with what you suggest. I consider this comment resolved.

> - All other `@id` elements from schemas, fields, and properties must be _under_ the root DTMI established by the interface.

Reviewer: This is not a DTDL @id requirement. It is legal to create a model with an @id for an element that is not "under" the interface's DTMI. The model repo could have this restriction, but it would mean that the model repo would not be able to store all possible models but only a subset of all possible models.

Author: that's our thinking. a restriction on this repo will be implemented by the PR validation checks.

Reviewer: The only service/tool that I know of that makes heavy use of @ids on elements other than Interface is Central. Does this restriction on @ids within an interface work with the models exported from Central?

Author: DTDL V1 Models exported from Central follow this pattern (same as the C# parser for system ids)

```json
{
  "@id": "urn:txs:Peripherals:2",
  "@type": "Interface",
  "displayName": {
    "en": "Peripherals"
  },
  "contents": [
    {
      "@id": "urn:txs:Peripherals:rt:1",
      "@type": "Telemetry",
      "displayName": {
        "en": "Runtime"
      },
      "name": "rt",
      "schema": "integer"
    },
    {
      "@id": "urn:txs:Peripherals:csq:1",
      "@type": "Telemetry",
      "displayName": {
        "en": "CSQ"
      },
```

I'm assuming v2 will keep this export format style

Reviewer: Okay, we should make sure Central knows this.

I consider this comment resolved.
